packages/protocol/contracts/common/MetaTransactionWallet.sol

```diff
@@ -0,0 +1,179 @@
+pragma solidity ^0.5.3;
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import "openzeppelin-solidity/contracts/utils/Address.sol";
+
+import "./interfaces/ICeloVersionedContract.sol";
+import "./interfaces/IMetaTransactionWallet.sol";
+import "./Initializable.sol";
+import "./Signatures.sol";
+
+contract MetaTransactionWallet is
+  IMetaTransactionWallet,
+  ICeloVersionedContract,
+  Initializable,
+  Ownable
+{
+  using SafeMath for uint256;
+
+  bytes32 public EIP172_DOMAIN_SEPARATOR;
+  // bytes32 public constant EIP172_EXECUTE_META_TRANSACTION_TYPEHASH = keccak256("ExecuteMetaTransaction(address destination,uint256 value,bytes data,uint256 nonce)");
+  bytes32 public constant EIP172_EXECUTE_META_TRANSACTION_TYPEHASH = 0x509c6e92324b7214543573524d0bb493d654d3410fa4f4937b3d2f4a903edd33;
+  uint256 public nonce;
+  address public signer;
+
+  event SignerSet(address signer);
+  event EIP172DomainSeparatorSet(bytes32 eip172DomainSeparator);
+  event TransactionExecution(address destination, uint256 value, bytes data, bytes returnData);
+  event MetaTransactionExecution(address destination, uint256 value, bytes data, bytes returnData);
+
+  /**
+    * @dev Fallback function allows to deposit ether.
+    */
+  function() external payable {}
+
+  /**
+   * @notice Returns the storage, major, minor, and patch version of the contract.
+   * @return The storage, major, minor, and patch version of the contract.
+   */
+  function getVersionNumber() public pure returns (uint256, uint256, uint256, uint256) {
+    return (1, 1, 0, 0);
+  }
+
+  /**
+   * @notice Used in place of the constructor to allow the contract to be upgradable via proxy.
+   * @param _signer The address authorized to execute transactions via this wallet.
+   */
+  function initialize(address _signer, uint256 chainId) external initializer {
+    setSigner(_signer);
+    setEip172DomainSeparator(chainId);
+    // MetaTransactionWallet owns itself, which necessitates that all onlyOwner functions
+    // be called via executeTransaction or executeMetaTransaction.
+    // If the signer was the owner, onlyOwner functions would not be callable via
+    // meta-transactions.
+    _transferOwnership(address(this));
+  }
+
+  /**
+   * @notice Transfers control of the wallet to a new signer.
+   * @param _signer The address authorized to execute transactions via this wallet.
+   */
+  function setSigner(address _signer) public onlyOwner {
```

Comment:
Couldn't the signer just execute this function directly as well?

Reply:
For posterity: They could, but per the comment above `_transferOwnership(this)` that would prevent `setSigner` from being callable via meta-transactions.